### ISSUE TITLE
Content Server "no books found"

### ISSUE THREAD

**Issue by beerygaz** [aptalca/docker-rdp-calibre#12]
i'm running the docker image on a Sinology NAS.  I can successfully connect to the GUI and have created a library, but whenever I try to access the content server, there are no books found.

Clicking on "All books" on the hoe page produces this error:

Error: No books found

TypeError: this.undef is not a function

    at Object.createException (http://nas.beerygaz.com:9091/static/stacktrace.js:81:18)

    at Object.run (http://nas.beerygaz.com:9091/static/stacktrace.js:66:25)

    at printStackTrace (http://nas.beerygaz.com:9091/static/stacktrace.js:57:62)

    at render_error (http://nas.beerygaz.com:9091/static/browse/browse.js:134:18)

    at booklist (http://nas.beerygaz.com:9091/static/browse/browse.js:271:29)

    at HTMLDocument. (http://nas.beerygaz.com:9091/browse/category/allbooks:34:17)

    at Function.ready (http://nas.beerygaz.com:9091/static/jquery.js:392:9)

    at HTMLDocument.DOMContentLoaded (http://nas.beerygaz.com:9091/static/jquery.js:745:10)

**Comment by aptalca** [comment]
It looks like you have some kind of proxy set up.

What happens if you go to the local IP of the server on the LAN?

**Comment by beerygaz** [comment]
No proxy.  The container is running on Synology NAS with port mapping on the NAS to the ports on the container network.

**Comment by aptalca** [comment]
See what happens on the local lan ip. 

If you get the same issue, please post your docker log and a screenshot of your container settings

**Comment by noinip** [comment]
Getting the same in unraid on a fresh install of the docker. I changed the host ports to 8888 and 8889 and set the library variable to /mnt/user/Media/Books. It said it couldn't start on port 8080 so I changed it to port 8081. Still said port was being used so I set it to port 8889 and I could get to the web server but I got the error above in OPs post


```
Tue Jul 26 22:22:47 2016
vncext: VNC extension running!
vncext: Listening for VNC connections on port 5901
vncext: created VNC server for screen 0
error opening security policy file /etc/X11/xserver/SecurityPolicy
Obt-Message: XKB extension is not present on the server or too old
Obt-Message: Xinerama extension is not present on the server
Xlib: extension "XInputExtension" missing on display ":1".
Xlib: extension "XInputExtension" missing on display ":1".
Fontconfig warning: ignoring C.UTF-8: not a valid language tag
Openbox-Message: Unable to find a valid menu file "/var/lib/openbox/debian-menu.xml"
Starting X app...
QXcbConnection: Failed to initialize XRandr
Qt: XKEYBOARD extension not present on the X server.
Fontconfig warning: ignoring C.UTF-8: not a valid language tag
22:22:48.137 [localhost-startStop-1] WARN o.g.g.properties.GuacamoleProperties - GuacamoleProperties is deprecated. Please use Environment instead.
22:22:48.143 [localhost-startStop-1] WARN o.g.g.properties.GuacamoleHome - GuacamoleHome is deprecated. Please use Environment instead.
22:22:48.145 [localhost-startStop-1] INFO o.g.g.n.b.r.a.BasicTokenSessionMap - Sessions will expire after 60 minutes of inactivity.
process 141: D-Bus library appears to be incorrectly set up; failed to read machine uuid: Failed to open "/etc/machine-id": No such file or directory
See the manual page for dbus-uuidgen to correct this issue.
22:22:48.446 [localhost-startStop-1] INFO o.g.g.n.b.w.t.WebSocketTunnelModule - Loading Tomcat 7 WebSocket support...
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.clipboard.ClipboardRESTService as a root resource class
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.connection.ConnectionRESTService as a root resource class
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.connectiongroup.ConnectionGroupRESTService as a root resource class
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.protocol.ProtocolRESTService as a root resource class
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.user.UserRESTService as a root resource class
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.auth.TokenRESTService as a root resource class
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.activeconnection.ActiveConnectionRESTService as a root resource class
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.codehaus.jackson.jaxrs.JacksonJsonProvider as a provider class
Jul 26, 2016 10:22:49 PM com.sun.jersey.server.impl.application.WebApplicationImpl _initiate
INFO: Initiating Jersey application, version 'Jersey: 1.17.1 02/28/2013 12:47 PM'
Jul 26, 2016 10:22:49 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.codehaus.jackson.jaxrs.JacksonJsonProvider to GuiceManagedComponentProvider with the scope "Singleton"
Jul 26, 2016 10:22:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.clipboard.ClipboardRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.connection.ConnectionRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.connectiongroup.ConnectionGroupRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.protocol.ProtocolRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.user.UserRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.auth.TokenRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:50 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.activeconnection.ActiveConnectionRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:50 PM com.sun.jersey.spi.inject.Errors processErrorMessages
WARNING: The following warnings have been detected with resource and/or provider classes:
WARNING: A sub-resource method, public void org.glyptodon.guacamole.net.basic.rest.activeconnection.ActiveConnectionRESTService.patchTunnels(java.lang.String,java.util.List) throws org.glyptodon.guacamole.GuacamoleException, with URI template, "/", is treated as a resource method
process 141: D-Bus library appears to be incorrectly set up; failed to read machine uuid: Failed to open "/etc/machine-id": No such file or directory
See the manual page for dbus-uuidgen to correct this issue.
DBusExport: Failed to connect to DBUS session bus, with error: org.freedesktop.DBus.Error.Spawn.ExecFailed: /usr/bin/dbus-launch terminated abnormally without any error message
Jul 26, 2016 10:22:50 PM org.apache.jasper.EmbeddedServletOptions <init>
SEVERE: The scratchDir you specified: /var/lib/tomcat7/work/Catalina/localhost/_ is unusable.
Jul 26, 2016 10:22:50 PM org.apache.catalina.startup.HostConfig deployWAR
INFO: Deploying web application archive /var/lib/tomcat7/webapps/guacamole.war
Jul 26, 2016 10:22:50 PM org.apache.catalina.core.StandardContext postWorkDirectory
WARNING: Failed to create work directory [/var/lib/tomcat7/work/Catalina/localhost/guacamole] for context [/guacamole]
process 141: D-Bus library appears to be incorrectly set up; failed to read machine uuid: Failed to open "/etc/machine-id": No such file or directory
See the manual page for dbus-uuidgen to correct this issue.
process 141: D-Bus library appears to be incorrectly set up; failed to read machine uuid: Failed to open "/etc/machine-id": No such file or directory
See the manual page for dbus-uuidgen to correct this issue.
22:22:53.671 [localhost-startStop-1] WARN o.g.g.properties.GuacamoleProperties - GuacamoleProperties is deprecated. Please use Environment instead.
22:22:53.677 [localhost-startStop-1] WARN o.g.g.properties.GuacamoleHome - GuacamoleHome is deprecated. Please use Environment instead.
22:22:53.678 [localhost-startStop-1] INFO o.g.g.n.b.r.a.BasicTokenSessionMap - Sessions will expire after 60 minutes of inactivity.
22:22:53.947 [localhost-startStop-1] INFO o.g.g.n.b.w.t.WebSocketTunnelModule - Loading Tomcat 7 WebSocket support...
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.clipboard.ClipboardRESTService as a root resource class
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.connection.ConnectionRESTService as a root resource class
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.connectiongroup.ConnectionGroupRESTService as a root resource class
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.protocol.ProtocolRESTService as a root resource class
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.user.UserRESTService as a root resource class
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.auth.TokenRESTService as a root resource class
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.glyptodon.guacamole.net.basic.rest.activeconnection.ActiveConnectionRESTService as a root resource class
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory register
INFO: Registering org.codehaus.jackson.jaxrs.JacksonJsonProvider as a provider class
Jul 26, 2016 10:22:54 PM com.sun.jersey.server.impl.application.WebApplicationImpl _initiate
INFO: Initiating Jersey application, version 'Jersey: 1.17.1 02/28/2013 12:47 PM'
Jul 26, 2016 10:22:54 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.codehaus.jackson.jaxrs.JacksonJsonProvider to GuiceManagedComponentProvider with the scope "Singleton"
Jul 26, 2016 10:22:55 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.clipboard.ClipboardRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:55 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.connection.ConnectionRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:55 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.connectiongroup.ConnectionGroupRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:55 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.protocol.ProtocolRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:55 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.user.UserRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:55 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.auth.TokenRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:55 PM com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory getComponentProvider
INFO: Binding org.glyptodon.guacamole.net.basic.rest.activeconnection.ActiveConnectionRESTService to GuiceManagedComponentProvider with the scope "PerRequest"
Jul 26, 2016 10:22:55 PM com.sun.jersey.spi.inject.Errors processErrorMessages
WARNING: The following warnings have been detected with resource and/or provider classes:
WARNING: A sub-resource method, public void org.glyptodon.guacamole.net.basic.rest.activeconnection.ActiveConnectionRESTService.patchTunnels(java.lang.String,java.util.List) throws org.glyptodon.guacamole.GuacamoleException, with URI template, "/", is treated as a resource method
Jul 26, 2016 10:22:55 PM org.apache.jasper.EmbeddedServletOptions <init>
SEVERE: The scratchDir you specified: /var/lib/tomcat7/work/Catalina/localhost/guacamole is unusable.
Jul 26, 2016 10:22:55 PM org.apache.coyote.AbstractProtocol start
INFO: Starting ProtocolHandler ["http-bio-8080"]
Jul 26, 2016 10:22:55 PM org.apache.catalina.startup.Catalina start
INFO: Server startup in 12477 ms
Jul 26, 2016 10:22:55 PM org.apache.coyote.AbstractProtocol start
INFO: Starting ProtocolHandler ["http-bio-8080"]
Jul 26, 2016 10:22:55 PM org.apache.catalina.startup.Catalina start
INFO: Server startup in 12477 ms
Traceback (most recent call last):
File "site-packages/calibre/library/server/base.py", line 276, in start
File "site-packages/calibre/library/server/base.py", line 269, in start_cherrypy
File "site-packages/cherrypy/process/wspbus.py", line 250, in start
ChannelFailures: IOError("Port 8081 not free on u'172.17.0.30'",)
```

**Comment by aptalca** [comment]
Sounds like you are using a custom library location and did not set it up properly. Post a screenshot of your container settings. 

The read me file here explains how to set it up properly

**Comment by noinip** [comment]
I reinstalled everything and kept everything vanilla. left the host/container ports at both 8080 and 8081. set /config to /mnt/cache/apps/appdata/calibre. Changed nothing else and I was still getting this error. I saw on the unraid forums you suggest to someone to point config to /mnt/cache/Programs/unRAID/appdata/Calibre/ and once I did that it started working. I'll play around more later...

**Comment by noinip** [comment]
Ok, further testing. I do want to have my own library path. First I setthe container volume to /library and then the host to /mnt/user/Media/Books/ and added it as read/write (this is in the unraid app menu) 
Then I added a new environment variable 'LIBRARYINTERNALPATH' and set it to /library

So far so good. I didn't actually move my library yet and I can see everything under the web server part. Everything being the Quick Start guide... 
Actually, not so good. The quick start guide is there but now it says "Unknown" for the title
Then I changed my library location to a brand new folder under that directory called test
There were no books so I added some. Showed up fine in Calibre gui but when I refreshed the web server I get the error in OPs post. I feel like I did all the steps you outline 2 separate times and I was very careful this last time:
First add a new mount point for the library location in the docker run command. Example: -v /path/to/library:/library:rw
Then add an environment variable (LIBRARYINTERNALPATH) to specify the internal library location for the webserver. Example: -e LIBRARYINTERNALPATH="/library"
When you fire up calibre the first time select your library location. Example: /library

I tried restarting the calibre-rdp container. Same issue as the original OP.  I then switched my library back to the one that was working and still had the same issue. I restarted the container and I had the same issue. Next I think I will try setting the config directory to my Media/Books

**Comment by noinip** [comment]
Does it normally create an directory with the name of a space in the config directory?
```
root@unRaid:/mnt/cache/Programs/unRAID/appdata/Calibre# ls
 
root@unRaid:/mnt/cache/Programs/unRAID/appdata/Calibre# cd \ /
root@unRaid:/mnt/cache/Programs/unRAID/appdata/Calibre/ # ls
John Schember  config  metadata.db
```

**Comment by noinip** [comment]
Apologies for all of the comments on this.

I've decided I get the error in OPs post no matter what if I don't initially set the config directory to /mnt/cache/Programs/unRAID/appdata/Calibre

**Comment by aptalca** [comment]
I'm sorry but this is pretty confusing and hard to follow. 
There should not be a directory with the name of a space. Please post a screenshot of your container settings. 
During the setup wizard on first run, you have to select the directory you set as the library internal path (in your case /library) otherwise the server interface won't work. 

The rdp gui and the webserver are two separate instances of calibre running. The gui uses whatever folder you select during the wizard as the library location. The webserver uses the library internal path location for it. Therefore those two have to be the same. If one is working and the other is not, then your settings are not correct. 

I would suggest starting over with a fresh install, empty config location, and set the library location to /config (defaults). Once you confirm that both the rdp and the webserver work through those ports, then you can delete the config folder and reinstall with the custom library location settings.

**Comment by noinip** [comment]
Apologies. Beer may have been involved.
Clean install....
I don't know how for two days I missed this

> IMPORTANT: On first start, select "/config" as the library location in the Calibre wizard

I am an idiot. Anyway, once I figured that out it solved my basic problem of not being able to use anything as the config folder other than /mnt/cache/Programs/unRAID/appdata/Calibre which must be whatever the default is that comes up when you first start to container.

I also was able to setup the LIBRARYINTERNALPATH without issue after this. I think I didn't understand that the /library has to point to a totally empty directory. I was trying to set the library up in the docker to like /library/calibre or whatever but the server portion was only looking in /library which doesn't have a library in it because I put it in /library/calibre. Anyway, it's working now. Hopefully this is a little easier to follow tonight :)

For OP. I'd make sure you setup the directory on first run of the container to just /config

For me, at least, I consider this issue solved.
